### PR TITLE
Generation Fetcher

### PR DESCRIPTION
**Pull request by macher259**
My take on [issue](https://github.com/piodul/scylla-cdc-rust/issues/2).
I have added some tests thanks to your idea from the slack, but I am not sure if I got what you meant right. 

Probably it's still WIP, but I would like to hear some feedback.

Tests compare output of the functions with output of the queries made in test tables that were populated with the data from my local database. 

Feel free to roast me.
Fixes: #2

### CODE REVIEW

**Review by piodul — CHANGES_REQUESTED** [review]
Some other issues to address:

- I think it would be nice to introduce a type which wraps all the necessary context for querying generation/stream info, and change the functions introduced here into methods of that type. You can capture the session object as `Arc<Session>`, and keep table names as `String`s.
- Try reducing repetition of magic numbers/strings in the test code. Ideally, stream IDs/timestamps/etc. should be given clear names and declared as constants at the beginning of the `tests` module.
- Before submitting a PR please make sure that it is cleanly split into commits. Each commit should be a separate, logical step, so that the reviewing process is easier. In this case, I think squashing everything into a single commit should be the right thing.
- When submitting a PR which fixes an issue, please add the following line at the end of the PR's description: `Fixes: #123`, where `#123` is the ID of the issue. By doing so, a link to the PR will appear on the issue's page, and the issue will be automatically closed when the PR is merged (and the merge commit will also contain the issue number, which might be useful when doing some git archaeology).

File: scylla-cdc/src/stream_generations.rs [hunk 1]
```diff
@@ -0,0 +1,407 @@
...
+    }
+
+    #[tokio::test]
+    async fn test_get_next_generation() {
+        let fetcher = setup().await.unwrap();
+
+        let gen = fetcher.fetch_all_generations().await.unwrap();
+
+        let gen1 = fetcher.fetch_next_generation(&gen[0]).await.unwrap();
+        assert!(gen1.is_none());
```
Comment (piodul) [inline scylla-cdc/src/stream_generations.rs:382]: It looks like the assertions are swapped in this test - the first invocation of `fetch_next_generation` should return the second generation, while second invocation of `fetch_next_generation` should return `None` because there is no more generations.

File: scylla-cdc/src/stream_generations.rs [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+mod tests {
+    use super::*;
+    use scylla::statement::Consistency;
+    use scylla::SessionBuilder;
+
+    const TEST_STREAM_TABLE: &str = "Test.cdc_streams_descriptions_v2";
+    const TEST_GENERATION_TABLE: &str = "Test.cdc_generation_timestamps";
+    const TEST_KEYSPACE: &str = "Test";
+    const GENERATION_1_MILLISECONDS: i64 = 1635882326384;
+    const GENERATION_2_MILLISECONDS: i64 = 1635882224341;
```
Comment (piodul) [inline scylla-cdc/src/stream_generations.rs:182]: The first generation's timestamp is larger than the second one. It would make more sense to be the other way around.

File: scylla-cdc/src/stream_generations.rs [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+    const TEST_STREAM_1: &str = "0x7fb9f781956cea08c651295720000001";
+    const TEST_STREAM_2: &str = "0x7fc0000000000000c298b9f168000001";
+    const TEST_UUID: &str = "d1dcf782-3d33-11ec-1b45-178f583c09c7";
```
Comment (piodul) [inline scylla-cdc/src/stream_generations.rs:185]: Honestly, it's hard to tell by looking at the raw UUID to which generation it is supposed to belong. It would be best to construct a timeuuid in such a way that it is clear by looking at the code to which generation it belongs - for example take a timestamp in the middle between `GENERATION_1_MILLISECONDS` and `GENERATION_2_MILLISECONDS` and construct a timeuuid from it, then check if it belongs to the first generation in a test.

Some documentation links which can be useful:
- Construct a uuid from timestamp: https://docs.rs/uuid/0.8.2/uuid/struct.Uuid.html#method.new_v1
- Construct a timestamp from seconds, nanoseconds and a ClockSequence: https://docs.rs/uuid/0.8.2/uuid/v1/struct.Timestamp.html#method.from_unix
- Construct a Context, which implements ClockSequence: https://docs.rs/uuid/0.8.2/uuid/v1/struct.Context.html#method.new

**Review by piodul — CHANGES_REQUESTED** [review]
There is a problem with the consistency level chosen for the queries. We would like the `GenerationFetcher` to use as high consistency level as possible, however the `system_distributed` keyspace has a replication factor of 3, but the cluster might have less than 3 nodes. Because of that it would be nice if, before each query, the `GenerationFetcher` checked the cluster size and then used appropriate consistency level:

- For 1 node use `Consistency::One`,
- For 2 nodes use `Consistency::Quorum`,
- For 3 nodes use `Consistency::All`.

To get the number of nodes you can send the following query: `SELECT COUNT(*) FROM system.peers`, then add 1 to the returned number. You can use the default consistency level for that query - the `system` keyspace is special and the coordinator which handles the query will only read local data and will not contact other nodes, so the query will work with any cluster size larger than 0.

File: scylla-cdc/src/stream_generations.rs [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+    "#,
+            self.generations_table_name
+        )
+    }
+
+    // Internal function configurable for testing purposes.
+    pub async fn fetch_generation_by_timeuuid(
+        &self,
+        time: &Uuid,
+    ) -> Result<Option<Generation>, Box<dyn Error>> {
```
Comment (Ponewor) [inline scylla-cdc/src/stream_generations.rs:80]: Does this really cover the requirement described as 

> For a given timestamp, which generation contains it

? I'm not sure if a timestamp is the same thing as an Uuid. Also, it kinda makes my work in #4 unfeasible.
@piodul what do you think?

Comment (piodul) [reply scylla-cdc/src/stream_generations.rs:80]: A timestamp is not the same as a timeuuid, but a timeuuid contains a timestamp + some other parts. This function does not use "the other parts" (because `dateOf` gets rid of them), and it's easier to convert a timeuuid to a timestamp, so I agree that this function should receive a timestamp, not a timeuuid.

**Comment by kbr-** [comment]
> For 3 nodes use Consistency::All.

We should be able to tolerate one failure, therefore we should use Quorum. It gives us the same consistency guarantees as All but has better liveness properties.

File: scylla-cdc/src/stream_generations.rs [hunk 5]
```diff
@@ -0,0 +1,407 @@
...
+    SELECT streams
+    FROM {}
+    WHERE time = ?;
+    "#,
+            self.streams_table_name
+        )
+    }
+
+    // Internal function configurable for testing purposes.
+    pub async fn fetch_stream_ids(
```
Comment (Ponewor) [inline scylla-cdc/src/stream_generations.rs:132]: This should work as well xD

```
pub async fn fetch_stream_ids(
        &self,
        timestamp: &Generation,
    ) -> Result<Vec<StreamID>, Box<dyn Error>> {
        Ok(self.session.query_iter(
            self.get_stream_ids_by_time_query(),
            (Timestamp(timestamp.0), ),
        ).await?.map(|x| StreamID(x.unwrap().into_typed::<(Vec<u8>, )>().unwrap().0)).collect().await)
    }

Comment (macher259) [reply scylla-cdc/src/stream_generations.rs:132]: I don't think so since each row contains `set<blob>` and cql set's equivalent is `Vec<T>` and cql blob's equivalent is `Vec<u8>`, so it ought to be `[...].into_typed::<(Vec<Vec<u8>>, )>()[...]`. I had some problem when trying to implement it, but it could be because of my lack of understanding of Rust.

Comment (Ponewor) [reply scylla-cdc/src/stream_generations.rs:132]: > each row contains `set<blob>`
Would you be so kind to point me to the specific part of documentation supporting this claim?

Comment (macher259) [reply scylla-cdc/src/stream_generations.rs:132]: The table we are querying is `system_distributed.cdc_streams_descriptions_v2`. Using cqlsh we can check its schema using `DESCRIBE`. It gives following output:
```
cqlsh> describe system_distributed.cdc_streams_descriptions_v2;

CREATE TABLE system_distributed.cdc_streams_descriptions_v2 (
    time timestamp,
    range_end bigint,
    streams frozen<set<blob>>,
    PRIMARY KEY (time, range_end)
) WITH CLUSTERING ORDER BY (range_end ASC)
    AND bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = ''
    AND compaction = {'class': 'SizeTieredCompactionStrategy'}
    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.0
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99.0PERCENTILE';
```

Comment (Ponewor) [reply scylla-cdc/src/stream_generations.rs:132]: Yeah, you're right, thank you.

Just for fun: this one works but it's admittedly messier ;)
```rust
pub async fn fetch_stream_ids(
        &self,
        timestamp: &Generation,
    ) -> Result<Vec<StreamID>, Box<dyn Error>> {
        Ok(self.session.query_iter(
            self.get_stream_ids_by_time_query(),
            (Timestamp(timestamp.0), ),
        ).await?.into_typed::<(Vec<Vec<u8>>, )>().flat_map(
            |x| futures::stream::iter(x.unwrap().0.iter().map(|y| StreamID(y.clone())).collect::<Vec<_>>())
        ).collect().await)
    }
```

File: scylla-cdc/src/stream_generations.rs [hunk 6]
```diff
@@ -0,0 +1,407 @@
...
+                Some(row) => {
+                    let row = row?.0;
+                    Ok(Some(Generation(row)))
+                }
+            };
+        }
+
+        Ok(None)
+    }
+}
```
Comment (Ponewor) [inline scylla-cdc/src/stream_generations.rs:170]: Variable shadowing is cool but isn't using one name for four different types a bit too much? At least it's a bit confusing to me, especially because initially what you're refferring to as a singular `row` is actually a `Vec<Row>` (and yeah, I know its length is always 1 in your use cases but it still gives the opposite impression). What do you think @macher259 ? Maybe I am wrong and it's exactly what you're supposed to do in Rust? You probably know better.

I like it this way:
```rust
pub fn return_single_row(rows: Option<Vec<Row>>) -> Result<Option<Generation>, Box<dyn Error>> {
        if let Some(rows) = rows {
            return match rows.into_typed::<(chrono::Duration, )>().next() {
                None => Ok(None),
                Some(row) => {
                    let time = row?.0;
                    Ok(Some(Generation(time)))
                }
            };
        }

        Ok(None)
    }
```

Comment (Ponewor) [reply scylla-cdc/src/stream_generations.rs:170]: As a side note: you can shorten this to
```rust
pub fn return_single_row(row: Option<Vec<Row>>) -> Result<Option<Generation>, Box<dyn Error>> {
    if let Some(row) = row {
        if let Some(row) = row.into_typed::<(chrono::Duration, )>().next() {
            let row = row?.0;
            return Ok(Some(Generation(row)))
        }
    }

    Ok(None)
}
```
